manager/src/main/scala/org/openinfralabs/caerus/cache/manager/BasicStorageIOMultiTierPlanner.scala

```diff
@@ -123,96 +124,82 @@ case class BasicStorageIOPlanner(optimizer: Optimizer, predictor: Predictor, pat
-
-    // Find initial cost for current contents.
-    val initialCost: Long = findCost(contents, plans)
-    logger.info("Initial I/O: %s MiB".format(initialCost /(1024*1024)))
-
-    // Find candidate per byte benefit.
-    val costs: Seq[Long] = remainingCandidates.map(candidate => {
-      findCost(contents + ((candidate,"temp")), plans) +
-        findSourceReadSize(contents, candidate) +
-        candidate.sizeInfo.get.writeSize
-    })
-    val benefits: Seq[Long] = costs.map(cost => initialCost - cost)
-    val benefitsPerByte: Seq[Long] = benefits.indices.map(i => {
-      benefits(i) / remainingCandidates(i).sizeInfo.get.writeSize
-    })
-    logger.info("Benefits/Byte: %s".format(benefitsPerByte.mkString(", ")))
-
-    val topCandidateIndex: Int = benefitsPerByte.zipWithIndex.maxBy(_._1)._2
-    val topCandidate = remainingCandidates(topCandidateIndex)
-    var newCost: Long = costs(topCandidateIndex)
-    if (newCost - initialCost >= 0) {
-      logger.info("No positive candidate found. Original plan is returned")
-      return optimizedPlan
-    }
-    logger.info("Top candidate: %s".format(topCandidate))
-
-    val topCandidateName: String = getName(topCandidate)
-    val topCandidatePath: String = path + Path.SEPARATOR + topCandidateName
-    var newContents = contents + ((topCandidate, topCandidatePath))
-    var newSize: Long = newContents.keys.map(candidate => candidate.sizeInfo.get.writeSize).sum
-    while (newSize > capacity) {
-      logger.warn("Capacity is not big enough to hold an extra candidate. Do not change anything.")
-      val bottomCandidateInfo: (Candidate, Long) = findBottomCandidate(newContents, plans, newCost)
-      val bottomCandidate: Candidate = bottomCandidateInfo._1
-      newCost = bottomCandidateInfo._2
-      newContents -= bottomCandidate
-      newSize -= bottomCandidate.sizeInfo.get.writeSize
-    }
-    logger.info("New contents:\n%s".format(newContents.mkString("\n")))
-
-    if (newContents.contains(topCandidate)) {
-      val caerusWrite: CaerusWrite = topCandidate match {
-        case Repartitioning(source, index, _) =>
-          CaerusRepartitioning(topCandidatePath, source, index)
-        case FileSkippingIndexing(source, index, _) =>
-          CaerusFileSkippingIndexing(topCandidatePath, source, index)
-        case Caching(cachedPlan, _) =>
-          val optimizedCachedPlan = optimizer.optimize(cachedPlan, newContents - topCandidate, emptyAddReference)
-          CaerusCaching(topCandidatePath, cachedPlan, optimizedCachedPlan)
-      }
-      val caerusDelete: Option[CaerusDelete] = if (!contents.keySet.subsetOf(newContents.keySet)) {
-        val deletedCandidates: Set[Candidate] = contents.keySet -- newContents.keySet
-        Some(CaerusDelete(deletedCandidates.toSeq.map(contents)))
-      } else {
-        None
+    var all_candidates: Seq[Candidate] = candidates
+    for(tier <- Tier.values){
+      if(all_contents.keySet.contains(tier)){
```

Comment:
if (allContents.contains(tier)) or if(tier in allContents) should be cheaper because they do not construct a new set.

Reply:
I didn't know you can directly check map's key, let me try